### PR TITLE
PML-141 refactor measurementstrategy to class based api with deprecation handling

### PR DESCRIPTION
**Pull request by CassNot**
## Summary

1. Implements the migration from `MeasurementStrategy` enum to factory‑based strategies, adds deprecation shims for legacy enum access, enforces `computation_space` ownership inside strategies, and wires `grouping` to run post‑measurement through the strategy pipeline.
2. Updates strategy processing signatures and adds tests that compare grouped outputs to legacy `MeasurementStrategy.PROBABILITIES`.
3. There is **no documentation update yet**

## Related Jira ticket (required)

Related Jira: PML-141

## Context / Related Issues

Completes the v0.3 measurement strategy refactor: enum values are deprecated, factory methods are the preferred API, and measurement configuration (computation_space + grouping) lives inside the strategy.

## Type of change
- [x] New feature
- [x] Refactor / Cleanup

## Proposed changes

- Migrate measurement strategies to factory‑based API (`probs`, `mode_expectations`, `partial`).
- Add deprecation warnings for legacy enum access and string strategies (removal in v0.4).
- Enforce `computation_space` resolution inside `MeasurementStrategy` and reject conflicting constructor usage.
- Apply `MeasurementStrategy.grouping `post‑measurement inside the strategy pipeline.
- Add tests validating grouping behavior vs legacy `MeasurementStrategy.PROBABILITIES`
- Add legacy alias [`MeasurementStrategy.NONE`] (non-deprecated) and map it to amplitudes.
- Update deprecation warnings to include “Will be removed in 0.4 (v0.4).” and clearer migration hints.
- Add tests for , `measurement_strategy=None` defaulting to probabilities, and new API backprop.
- Add CUDA-guarded tests for device correctness (probabilities + NONE).

```
from merlin import MeasurementStrategy, ComputationSpace, QuantumLayer, LexGrouping

layer = QuantumLayer(
    input_size=3,
    n_photons=2,
    builder=builder,
    measurement_strategy=MeasurementStrategy.probs(
        computation_space=ComputationSpace.UNBUNCHED,
        grouping=LexGrouping(6,2)
    ),
)

```
## How to test / How to run

1. Command lines

```
pytest tests/measurement/test_measurement_strategy.py -k grouping
pytest tests/measurement/test_measurement_strategy.py 
pytest 
```

## Documentation
- [x] Docstrings updated

## Checklist
- [x] PR title includes Jira issue key (e.g., PML-126)
- [x] "Related Jira ticket" section includes the Jira issue key (no URL)
- [x] Code formatted (ruff format)
- [x] Lint passes (ruff)
- [x] Static typing passes (mypy) if applicable
- [x] Unit tests added/updated (pytest)
- [x] Tests pass locally (pytest)
- [x] Tests pass on GPU (pytest)
- [ ] Test coverage not decreased significantly
- [ ] Docs build locally if affected (sphinx)
- [x] Dependencies updated (if needed) and pinned appropriately
- [x] PR description explains what changed and how to validate it

### CODE REVIEW

File: merlin/algorithms/feed_forward.py [hunk 1]
```diff
@@ -113,9 +117,9 @@ class FeedForwardBlock(MerlinModule):
     computation_space:
         Currently restricted to :attr:`~merlin.core.computation_space.ComputationSpace.FOCK`.
     measurement_strategy:
-        Controls how classical outputs are produced:
+        Controls how classical outputs are produced.
 
-        - ``MeasurementStrategy.PROBABILITIES`` (default) returns a tensor of
+        - ``MeasurementStrategy.PROBABILITIES`` returns a tensor of
```
Comment (philippeschoeb) [inline merlin/algorithms/feed_forward.py]: PROBABILITIES is still the default option. I think it is great to indicate it

File: merlin/algorithms/feed_forward.py [hunk 2]
```diff
@@ -497,7 +503,7 @@ def _build_stage_runtime(
                 n_photons=self.n_photons,
                 trainable_parameters=trainable_parameters,
                 input_parameters=input_parameters,
-                measurement_strategy=MeasurementStrategy.AMPLITUDES,
+                measurement_strategy=MeasurementKind.AMPLITUDES,
```
Comment (philippeschoeb) [inline merlin/algorithms/feed_forward.py]: I am not sure why we don't use our MeasurementStrategy.none() / MeasurementStrategy.amplitudes() object here. I think it would be simpler than introducing MeasurmentKind

File: merlin/algorithms/layer.py [hunk 3]
```diff
@@ -162,12 +168,14 @@ def __init__(
...
-            "dual_rail"}``. If omitted, defaults to ``UNBUNCHED``.
-        measurement_strategy : MeasurementStrategy, default: PROBABILITIES
-            Output mapping strategy. Supported values include ``PROBABILITIES``,
-            ``MODE_EXPECTATIONS`` and ``AMPLITUDES``.
+            "dual_rail"}``. If omitted, defaults to ``UNBUNCHED``. This argument
+            is deprecated; move it into ``MeasurementStrategy.probs(...)``.
+        measurement_strategy : MeasurementStrategy | None, default: None
+            Output mapping strategy. When omitted, defaults to
+            ``MeasurementStrategy.probs(computation_space)``. Supported values
+            include ``PROBABILITIES``, ``MODE_EXPECTATIONS`` and ``AMPLITUDES``.
```
Comment (philippeschoeb) [inline merlin/algorithms/layer.py]: I think `MeasurementStrategy.mode_expectations()`, `MeasurementStrategy.amplitudes()` should also be indicated here since they are the new supported usage

File: merlin/measurement/strategies.py [hunk 4]
```diff
@@ -113,49 +141,87 @@ def process(
...
-        if apply_sampling:
+        grouping: Callable[[torch.Tensor], torch.Tensor] | None = None,
+    ) -> PartialMeasurement:
+        if apply_sampling and effective_shots > 0:
             raise RuntimeError(
-                "Sampling cannot be applied when measurement_strategy=MeasurementStrategy.AMPLITUDES."
+                "Sampling cannot be applied when measurement_strategy=MeasurementStrategy.partial()."
             )
-        return amplitudes
+        detector_output = apply_detectors(amplitudes)
```
[the remark this answers is not in the document]
Comment (CassNot) [reply merlin/measurement/strategies.py]: in `layer.py`
I didn’t touch `strategies.py` because the cleanest fix is to block the unsupported combination at setup time.

**Review by philippeschoeb — COMMENTED** [review]
This is a first review to engage discussion (not to approve or request changes). 

My general feeling is that there are **too many "measurement util" classes**, that it makes the code harder to understand and harder to change for future versions. For instance, I think that the MeasurementStrategyLike class could be removed and I feel like a measurement_strategy should only be initializable with `MeasurmentStrategy.STRAT` (old enum version) or with `MeasurmentStrategy.strat()` (new method version) instead of including MeasurementKind in the mix (although I understand that the MeasurementKind class is necessary to distinguish measurement strategies).

Nevertheless, as I have not delved deep enough in the refactor of the MeasurementStrategy, there certainly are issues that I am not considering when making my suggestions (which is why I would like to discuss about it). So **you should interpret my comments as comments rather than as requested changes**.

**Comment by CassNot** [comment]
I think you are partially right:
- Right now the “too many measurement util classes” are mostly a side‑effect of supporting multiple entry points (**legacy** enum, new dataclass, plus `MeasurementKind`), and that support is leaking into runtime behaviour in a way that could be easy to misuse.
- Therefore, the public input needs to be changed because `MeasurementKind` could be misused and bypassed deprecation warnings. This includes `QuantumLayer.__init__` which documents` MeasurementStrategy | None`, but accepts `MeasurementStrategyLike`, which currently includes `MeasurementKind` and the legacy enum.

File: merlin/algorithms/feed_forward.py [hunk 5]
```diff
@@ -1113,16 +1128,19 @@ def _branches_to_classical(
...
+                == MeasurementKind.MODE_EXPECTATIONS
+            ):
                 return torch.zeros(
                     (0, self.total_modes), dtype=self.dtype, device=self.device
                 )
             return torch.zeros(0, device=self.device)
 
         strategy = self.measurement_strategy
-        if strategy == MeasurementStrategy.PROBABILITIES:
+        if _resolve_measurement_kind(strategy) == MeasurementKind["PROBABILITIES"]:
```
Comment (philippeschoeb) [inline merlin/algorithms/feed_forward.py]: Replace right side of == by `MeasurementKind.PROBABILITIES` for consistency with other MeasurementKind usage.

File: merlin/measurement/mappers.py [hunk 6]
```diff
@@ -67,15 +71,19 @@ def create_mapping(
         Raises:
             ValueError: If strategy is unknown
         """
-        if strategy == MeasurementStrategy.PROBABILITIES:
+        try:
+            kind = _resolve_measurement_kind(strategy)
+        except TypeError as exc:
+            raise ValueError(f"Unknown measurement strategy: {strategy}") from exc
+        if kind == MeasurementKind["PROBABILITIES"]:
```
Comment (philippeschoeb) [inline merlin/measurement/mappers.py]: Uniformize syntax: `MeasurementKind.PROBABILITIES` instead of `MeasurementKind["PROBABILITIES"]`

**Review by philippeschoeb — CHANGES_REQUESTED** [review]
Some small syntax inconsistencies, doc changes and feature changes to make before merge

File: merlin/algorithms/layer.py [hunk 7]
```diff
@@ -873,6 +897,12 @@ def forward(
 
         # Phase 6: Measurement strategy dispatch and output mapping
         strategy = resolve_measurement_strategy(self.measurement_strategy)
+        # Handle backward compatibility for backpropagation - will be removed in future
+        grouping = None
+        if isinstance(self.measurement_strategy, MeasurementStrategy):
+            if self.measurement_strategy.type == MeasurementKind.PROBABILITIES:
```
[the remark this answers is not in the document]
Comment (philippeschoeb) [reply merlin/algorithms/layer.py]: I think if the user specifies a grouping, it would be normal that the dimension of the probs change hence they would not fit with the number of StateVector. Plus, the support of grouping is already implemented in my PR to come for PML-146: see `Details about set_grouping()` in the PML-146 ticket for more details

**Review by philippeschoeb — APPROVED** [review]
